internal/impl/generic/cork.go

```diff
@@ -0,0 +1,287 @@
+			return
+		}
+	}
+}
+
+func (ci *corkInput) cork() {
+	if atomic.CompareAndSwapInt32(ci.status, modeUncorked, modeCorked) {
+		ci.logger.Debug("corking input")
+		ci.wg.Add(1)
+	}
+}
+
+func (ci *corkInput) uncork() {
+	if atomic.CompareAndSwapInt32(ci.status, modeCorked, modeUncorked) {
+		ci.logger.Debug("uncorking input")
+		ci.wg.Done()
+	}
+}
+
+func newCorkFromConfig(conf *service.ParsedConfig, logger *service.Logger) (service.BatchInput, error) {
+	wrapped, err := conf.FieldInput("input")
+	if err != nil {
+		return nil, err
+	}
+
+	signal, err := conf.FieldInput("signal")
+	if err != nil {
+		return nil, err
+	}
+
+	initCorked, err := conf.FieldBool("initially_corked")
+	if err != nil {
+		return nil, err
+	}
+
+	return newCorkInput(&CorkConfig{
+		Input:           wrapped,
+		Signal:          signal,
+		InitiallyCorked: initCorked,
+		Logger:          logger,
+	}), nil
+}
+
+func newCorkConfigSpec() *service.ConfigSpec {
+	return service.NewConfigSpec().
+		Beta().
+		Summary("Use signals from one input to suspend or resume reads from another input").
+		Categories("Utility").
+		Description(`
+This input operates in of two modes: corked and uncorked.
+When corked, reading from the underlying input is suspended.
+When uncorked, reading messages from the underlying input works as normal.
+
+When there are no more messages on the signal input, this input will continue running with the last mode that was set.
+`).
+		Field(service.NewBoolField("initially_corked").
+			Default(false).
+			Description("Sets the cork mode at the start of the process"),
+		).
+		Field(service.NewInputField("input").
+			Description("The input to wrap with a cork"),
+		).
+		Field(service.NewInputField("signal").
+			Description(`
+An input that supplies cork/uncork signals.
+
+It must produce one of two string values that control the read modes:
+
+- ` + "`\"cork\"`" + `: suspends reading from the underlying input
+- ` + "`\"uncork\"`" + `: resumes reading from the underlying input
+
+Any other values are ignored.
+`).
+			Example(map[string]interface{}{
+				"broker": map[string]interface{}{
+					"inputs": []interface{}{
+						map[string]interface{}{
+							"http_server": map[string]interface{}{
+								"path":          "/cork",
+								"allowed_verbs": []string{"POST"},
+							},
+							"processors": []interface{}{
+								map[string]interface{}{"bloblang": "root = \"cork\""},
+							},
+						},
+						map[string]interface{}{
+							"http_server": map[string]interface{}{
+								"path":          "/uncork",
+								"allowed_verbs": []string{"POST"},
+							},
+							"processors": []interface{}{
+								map[string]interface{}{"bloblang": "root = \"uncork\""},
+							},
+						},
+					},
+				},
+			}),
+		)
+}
+
+func init() {
+	err := service.RegisterBatchInput(
+		"cork",
+		newCorkConfigSpec(),
+		func(conf *service.ParsedConfig, mgr *service.Resources) (service.BatchInput, error) {
+			return newCorkFromConfig(conf, mgr.Logger())
```

Comment:
I decided not to add auto retry to this input as I wanted to delegate this concern to the underlying input.

Reply:
Yeah that's the right call